### DISCUSSION TITLE
multi-client telnet log

### DISCUSSION THREAD

**Discussion by RafSchandevyl** [eModbus/eModbus#299] [General]
Hi,

I'm using the RemoteDebug library in my application, and would like to redirect the log messages to this telnet debug.  How can ik Reassing the logdevice to this library ?

mvg
Raf

**Comment by Miq1** [comment]
Have a look at [the documentation](https://emodbus.github.io/logging#output-channel) - any object derived from ``Print`` will  do.

**Comment by RafSchandevyl** [comment]
Ok,

But how and where can I assign it to the LOGDEVICE ?    In the doc you are talking about a multi-client telnet example.  Could you provide this as a sample ?

Regards
Raf

**Reply by Miq1** [reply]
> Ok,
> 
> But how and where can I assign it to the LOGDEVICE ? 
In ``setup()`` for instance?

> In the doc you are talking about a multi-client telnet example. Could you provide this as a sample ?
> 
Have a look at the (monstrous, I know) [test sketch](https://github.com/eModbus/eModbus/blob/master/Test/main.cpp). 
For testing purposes I will create an alternative log device at line 22ff:
```
// Define alternate LOGDEVICE class to test rerouting of the log output
class AltLog : public Print {
  size_t write(uint8_t c) {
    Serial.write(c);
    return 1;
  }
  size_t write(const uint8_t *buffer, size_t size) {
    Serial.write(buffer, size);
    return size;
  }
};
AltLog a;
```
At line 469 in ``setup()`` I do a ``LOGDEVICE = &a;`` - that does the job.

The telnet log server mentioned you can find [here](https://github.com/eModbus/Smartdose/tree/main/src) - it is an async implementation for the ESP8266, so you might to have to change headers for the ESP32.
